baremetal/metal3data_manager.go

```diff
@@ -523,25 +581,43 @@ func (m *DataManager) getAddressFromM3Pool(ctx context.Context, poolRef corev1.T
 	}, false, nil
 }
 
-// releaseAddressFromM3Pool deletes the IP claim for a referenced pool.
+// releaseAddressFromM3Pool deletes the Metal3IPClaim for a referenced pool.
 func (m *DataManager) releaseAddressFromM3Pool(ctx context.Context, poolRef corev1.TypedLocalObjectReference) error {
-	ipClaim, err := fetchM3IPClaim(ctx, m.client, m.Log, m.Data.Name+"-"+poolRef.Name,
-		m.Data.Namespace,
-	)
+	var ipClaim *ipamv1.IPClaim
+	var err, finalizerErr error
+	IPClaimsList, err := m.fetchIPClaimsWithLabels(ctx, poolRef.Name)
+	if err == nil {
+		for _, IPClaimWithLabels := range IPClaimsList {
```

Comment:
> This looks like public/exported variables with the capital "IP" to me. Maybe better to use `ipClaimsList` and `ipClaimWithLabels` to keep it similar to `ipClaim` and also show that it is private.

Yes, actually I have to change that to make it aligned. 



> Also, if err != nil we just continue, is this expected? A comment to explain would be good in that case. If I understand correctly this part with the labels is for preallocations, so it makes sense that if we don't have any we continue to check the "normal" case. But what if there was some temporary network issue? Should we not requeue then instead?

yes, we do try first to fetch and delete the claims with preallocation labels regardless or not we are using this feature (to prevent leaving up the resources in case the flag for enabling the feature was changed from ON to OFF in the bad timing let's say). As is we continue if we do not find any object to delete, but I do not know if we should requeue actually at least we do not do it in most of the places in the codebase.

Reply:
> yes, we do try first to fetch and delete the claims with preallocation labels regardless or not we are using this feature (to prevent leaving up the resources in case the flag for enabling the feature was changed from ON to OFF in the bad timing let's say).

:+1:

> As is we continue if we do not find any object to delete, but I do not know if we should requeue actually at least we do not do it in most of the places in the codebase.

How about checking if we get a "not found" error? In that case it is at least clear that we can continue without issue. If there is some other error I would suggest at least logging it, not sure if we need to requeue.

Reply:
@lentzi90 deleteObject https://github.com/Nordix/cluster-api-provider-metal3/blob/9863d8f2805708088eb47866e877a78e75e00fe8/baremetal/utils.go#L112 already checks for "not found" error and returns nil if it meets that condition, otherwise returns error. What I did now is, I removed not found err check since that is part of the func already and return err when there is !not found error

Reply:
> Yes, actually I have to change that to make it aligned.

Done

Reply:
Thanks, LGTM!